File: other/c/fib.c
#include <stdio.h>

#include "common.h"

// DESCRIPTION
//   fib - find the nth fibonacci number
// SYNOPSIS
//   fib n
// BUILDING
//   $ meson build
//   $ meson compile -C build

// Note that this function uses 0-based indexing for the Fibonacci numbers.
unsigned long fib(long n) {
  unsigned long a = 0;
  unsigned long b = 1;

  for (long i = 0; i < n; ++i) {
    unsigned long next = a + b;

    a = b;
    b = next;
  }

  return a;
}

int main(int argc, char *argv[]) {
  if (argc == 1) {
    fprintf(stderr, "Please specify a number.\n");
    return 1;
  }

  long n = parse_long(argv[1]);
  if (n < 1) {
    fprintf(stderr, "The input number must be 1 or greater.\n");
    return 1;
  }
  printf("Fibonacci number #%lu is %lu.\n", n, fib(n - 1));

  return 0;
}

File: other/haskell/fib.hs
{-# LANGUAGE BangPatterns #-}

module Main where

import System.Environment
import System.Exit

-- DESCRIPTION
--   fib - find the nth fibonacci number
-- SYNOPSIS
--   fib n
-- BUILDING
--   $ ghc -dynamic fib.hs

main :: IO ()
main = getArgs >>= parse >>= run

parse :: [String] -> IO Int
parse (n : _) = do
  let n' = read n
  if n' >= 1
    then return n'
    else die "The input must be 1 or greater."
parse _ = die "Please specify a number."

run :: Int -> IO ()
run n = putStrLn $ "Fibonacci number #" ++ show n ++ " is " ++ result ++ "."
 where
  result = show $ fib (n - 1)

fib :: Int -> Int
fib n = go n (0, 1)
 where
  go !n (!a, !b)
    | n == 0 = a
    | otherwise = go (n - 1) (b, a + b)

-- There is also this iterative implementation:
-- fib :: Int -> Int
-- fib = (fibs !!)
--   where
--     fibs = 0 : 1 : zipWith (+) fibs (tail fibs)
